File: ttt_go/ttt/board.go
package ttt

import (
	"errors"
	"fmt"
)

/// Player 1 and 2.
const (
	Player1 int = 1
	Player2 int = 2
)

/// State of current game.
const (
	Playing = iota
	Winner  = iota
	Draw    = iota
)

/// Constants.
const (
	BoardSize int = 3
	TotalSize int = BoardSize * BoardSize
)

/// Board data.
type Board struct {
	// move first
	MoveFirst int

	// current
	Current int

	// history moves
	History string

	// move list
	MoveList [9]int

	// state
	State int
}

// ----------------------------------------------------------------------

/// Move current marker to board.
func (board *Board) Move(index int) bool {
	if !(index >= 0 && index < TotalSize) {
		return false
	}

	// check with MoveList to see it already move there
	if board.MoveList[index] > 0 {
		return false
	}

	// add move
	board.History += string('0' + rune(index))
	board.MoveList[index] = board.Current

	// check move result
	if board.CheckVictoryState() {
		board.State = Winner
	} else if board.CheckDrawState() {
		board.State = Draw
	} else {
		// next player
		if board.Current == Player1 {
			board.Current = Player2
		} else {
			board.Current = Player1
		}
	}

	return true
}

/// Finish move and can check result in State.
func (board *Board) Finish() bool {
	return board.State != Playing
}

// ----------------------------------------------------------------------

func (board *Board) CheckYVictory(y int) bool {
	x := 0
	for x < BoardSize {
		if board.MoveList[y*BoardSize+x] != board.Current {
			return false
		}
		x++
	}
	fmt.Printf("Player %v win with y=%v\n", board.Current, y)
	return true
}

func (board *Board) CheckXVictory(x int) bool {
	y := 0
	for y < BoardSize {
		if board.MoveList[y*BoardSize+x] != board.Current {
			return false
		}
		y++
	}
	fmt.Printf("Player %v win with x=%v\n", board.Current, x)
	return true
}

func (board *Board) CheckDiagonalVictory0() bool {
	i := 0
	for i < BoardSize {
		if board.MoveList[i*BoardSize+i] != board.Current {
			return false
		}
		i++
	}
	fmt.Printf("Player %v win with direct diagonal\n", board.Current)
	return true
}

func (board *Board) CheckDiagonalVictory1() bool {
	i := 0
	for i < BoardSize {
		if board.MoveList[i*BoardSize+(BoardSize-i-1)] != board.Current {
			return false
		}
		i++
	}
	fmt.Printf("Player %v win with reverse diagonal\n", board.Current)
	return true
}

// ----------------------------------------------------------------------

/// Check victory.
func (board *Board) CheckVictoryState() bool {
	// check horizontal
	for y := 0; y < BoardSize; y++ {
		if board.CheckYVictory(y) {
			return true
		}
	}

	// check vertical
	for x := 0; x < BoardSize; x++ {
		if board.CheckXVictory(x) {
			return true
		}
	}

	// check diagonal
	if board.CheckDiagonalVictory0() {
		return true
	}
	if board.CheckDiagonalVictory1() {
		return true
	}

	return false
}

/// Check draw.
func (board *Board) CheckDrawState() bool {
	for i := range board.MoveList {
		if board.MoveList[i] <= 0 {
			return false
		}
	}
	return true
}

// ----------------------------------------------------------------------

/// Create new board.
func New(moveFirst int) (*Board, error) {
	if !(moveFirst >= Player1 && moveFirst <= Player2) {
		return nil, errors.New("parameter moveFirst must be Player1 or Player2")
	}

	this := Board{
		MoveFirst: moveFirst,
		Current:   moveFirst,
		History:   "",
		MoveList:  [9]int{0, 0, 0, 0, 0, 0, 0, 0, 0},
		State:     Playing,
	}
	return &this, nil
}

/// Load board from history.
func Load(moveFirst int, history string) (*Board, error) {
	if !(moveFirst >= Player1 && moveFirst <= Player2) {
		return nil, errors.New("parameter moveFirst must be Player1 or Player2")
	}

	p := moveFirst
	r := [9]int{0, 0, 0, 0, 0, 0, 0, 0, 0}
	l := len(history)
	for i := 0; i < l; i++ {
		c := history[i]
		if !(c >= '0' && c < '9') {
			return nil, errors.New("found unexpected character in history, only 0..8 allow")
		}

		// extract history and copy into MoveList
		v := int(c - '0')
		r[v] = p

		// next player
		if p == Player1 {
			p = Player2
		} else {
			p = Player1
		}
	}

	this := Board{
		MoveFirst: moveFirst,
		Current:   moveFirst,
		History:   history,
		MoveList:  r,
		State:     Playing,
	}

	if this.CheckVictoryState() || this.CheckDrawState() {
		return nil, errors.New("this game result is concluded")
	}

	return &this, nil
}

File: ttt_rust/play/src/board.rs
/// Player 1 or 2.
pub const PLAYER_1: u8 = 1;
pub const PLAYER_2: u8 = 2;

/// State of current game.
#[derive(PartialEq)]
pub enum State {
	PLAYING = 0,
	WINNER = 1,
	DRAW = 2,
}

/// Other constants.
pub const BOARD_SIZE: u8 = 3;
pub const TOTAL_SIZE: u8 = BOARD_SIZE * BOARD_SIZE;

/// Board data.
pub struct Board {
	// move first
	pub move_first: u8,

	// current
	pub current: u8,

	// history moves
	pub history: String,

	// move list
	pub move_list: [u8; 9],

	// state
	pub state: State,
}

// ----------------------------------------------------------------------

/// Board implementation.
impl Board {
	/// Move current marker to board.
	pub fn move_to(&mut self, index: u8) -> bool {
		if !(index < TOTAL_SIZE) {
			return false;
		}

		// check with move_list to see it already move there
		if self.move_list[index as usize] > 0 {
			return false;
		}

		// add move
		self.history.push_str(&index.to_string());
		self.move_list[index as usize] = self.current as u8;

		// check move result
		if self.check_victory_state() {
			self.state = State::WINNER
		} else if self.check_draw_state() {
			self.state = State::DRAW
		} else {
			// next player
			if self.current == PLAYER_1 {
				self.current = PLAYER_2
			} else {
				self.current = PLAYER_1
			}
		}

		return true;
	}

	/// Finish move and can check result in State.
	pub fn is_finished(&self) -> bool {
		return self.state != State::PLAYING;
	}

	// ----------------------------------------------------------------------

	fn check_y_victory(&self, y: u8) -> bool {
		let mut x = 0;
		while x < BOARD_SIZE {
			let u = (y * BOARD_SIZE + x) as usize;
			if self.move_list[u] != self.current {
				return false;
			}
			x = x + 1;
		}
		println!("Player {} win with y={}", self.current, y);
		return true;
	}

	fn check_x_victory(&self, x: u8) -> bool {
		let mut y = 0;
		while y < BOARD_SIZE {
			let u = (y * BOARD_SIZE + x) as usize;
			if self.move_list[u] != self.current {
				return false;
			}
			y = y + 1;
		}
		println!("Player {} win with x={}", self.current, x);
		return true;
	}

	fn check_diagonal_victory(&self) -> bool {
		let mut i = 0;
		while i < BOARD_SIZE {
			let u = (i * BOARD_SIZE + i) as usize;
			if self.move_list[u] != self.current {
				return false;
			}
			i = i + 1;
		}
		println!("Player {} win with direct diagonal", self.current);
		return true;
	}

	fn check_diagonal_victory_reserse(&self) -> bool {
		let mut i = 0;
		while i < BOARD_SIZE {
			let u = (i * BOARD_SIZE + (BOARD_SIZE - i - 1)) as usize;
			if self.move_list[u] != self.current {
				return false;
			}
			i = i + 1;
		}
		println!("Player {} win with reverse diagonal", self.current);
		return true;
	}

	// ----------------------------------------------------------------------

	/// Check victory.
	pub fn check_victory_state(&self) -> bool {
		// check horizontal
		let mut y = 0;
		while y < BOARD_SIZE {
			if self.check_y_victory(y) {
				return true;
			}
			y = y + 1;
		}

		// check vertical
		let mut x = 0;
		while x < BOARD_SIZE {
			if self.check_x_victory(x) {
				return true;
			}
			x = x + 1;
		}

		// check diagonal
		if self.check_diagonal_victory() {
			return true;
		}
		if self.check_diagonal_victory_reserse() {
			return true;
		}

		return false;
	}

	/// Check draw.
	pub fn check_draw_state(&self) -> bool {
		for i in self.move_list {
			if i <= 0 {
				return false;
			}
		}
		return true;
	}
}

// ----------------------------------------------------------------------

pub fn new(move_first: u8) -> Board {
	if !(move_first >= PLAYER_1 && move_first <= PLAYER_2) {
		panic!("parameter moveFirst must be Player1 or Player2");
	}

	Board {
		move_first,
		current: move_first,
		history: String::from(""),
		move_list: [0, 0, 0, 0, 0, 0, 0, 0, 0],
		state: State::PLAYING,
	}
}
